ckanext/dcat/harvesters/_json.py

```diff
@@ -252,35 +252,39 @@ def import_stage(self, harvest_object):
             'ignore_auth': True,
         }
 
-        if status == 'new':
-
-            package_schema = logic.schema.default_create_package_schema()
-            context['schema'] = package_schema
-
-            # We need to explicitly provide a package ID
-            package_dict['id'] = unicode(uuid.uuid4())
-            package_schema['id'] = [unicode]
-
-            # Save reference to the package on the object
-            harvest_object.package_id = package_dict['id']
-            harvest_object.add()
-
-            # Defer constraints and flush so the dataset can be indexed with
-            # the harvest object id (on the after_show hook from the harvester
-            # plugin)
-            model.Session.execute(
-                'SET CONSTRAINTS harvest_object_package_id_fkey DEFERRED')
-            model.Session.flush()
-
-            package_id = \
-                p.toolkit.get_action('package_create')(context, package_dict)
-            log.info('Created dataset with id %s', package_id)
-
-        elif status == 'change':
-            package_dict['id'] = harvest_object.package_id
-            package_id = \
-                p.toolkit.get_action('package_update')(context, package_dict)
-            log.info('Updated dataset with id %s', package_id)
+        try:
+            if status == 'new':
+
+                package_schema = logic.schema.default_create_package_schema()
+                context['schema'] = package_schema
+
+                # We need to explicitly provide a package ID
+                package_dict['id'] = unicode(uuid.uuid4())
+                package_schema['id'] = [unicode]
+
+                # Save reference to the package on the object
+                harvest_object.package_id = package_dict['id']
+                harvest_object.add()
+
+                # Defer constraints and flush so the dataset can be indexed with
+                # the harvest object id (on the after_show hook from the harvester
+                # plugin)
+                model.Session.execute(
+                    'SET CONSTRAINTS harvest_object_package_id_fkey DEFERRED')
+                model.Session.flush()
+
+                package_id = \
+                    p.toolkit.get_action('package_create')(context, package_dict)
+                log.info('Created dataset with id %s', package_id)
+
+            elif status == 'change':
+                package_dict['id'] = harvest_object.package_id
+                package_id = \
+                    p.toolkit.get_action('package_update')(context, package_dict)
+                log.info('Updated dataset with id %s', package_id)
+        except p.toolkit.ValidationError, e:
```

Comment:
Do we know if this exception can be raised on any line within the `try` block? I wonder if we can get the exception handling closer to where it's likely to occur?

Reply:
Could do, I just wanted to cover the 2 different scenarios which could throw it but I think I might have a way of refactoring it.